python/sqlflow_submitter/xgboost/train.py

```diff
@@ -44,14 +44,15 @@ def train(datasource,
                          cache=disk_cache,
                          batch_size=batch_size,
                          epoch=epoch)
+    if len(validation_select.strip()) > 0:
+        dvalidate = list(
+            xgb_dataset(datasource, 'validate.txt', validation_select,
+                        feature_metas, feature_column_names, label_meta,
+                        is_pai, pai_validate_table))[0]
     bst = None
     for per_batch_dmatrix in dtrain:
         watchlist = [(per_batch_dmatrix, "train")]
         if len(validation_select.strip()) > 0:
-            dvalidate = list(
-                xgb_dataset(datasource, 'validate.txt', validation_select,
-                            feature_metas, feature_column_names, label_meta,
-                            is_pai, pai_validate_table))[0]
             watchlist.append((dvalidate, "validate"))
```

Comment:
move L56 to L51?

Reply:
Must construct the `watchlist` ever iteration. If move this like up, it should look like:

```python
if len(validation_select.strip()) > 0:
        dvalidate = list(
            xgb_dataset(datasource, 'validate.txt', validation_select,
                        feature_metas, feature_column_names, label_meta,
                        is_pai, pai_validate_table))[0]
        watchlist = [(dvalidate, "validate")]
    else:
        watchlist = []
    bst = None
    for per_batch_dmatrix in dtrain:
        if len(validation_select.strip()) == 0:
            watchlist = [(per_batch_dmatrix, "train")]
        else:
            if step == 0:
                watchlist.append((per_batch_dmatrix, "train"))
            else:
                watchlist[1] = (per_batch_dmatrix, "train")
```
it's too complicated.

Reply:
I see, you're right.